store/src/sql/access/event.rs

```diff
@@ -61,6 +123,56 @@ impl CeramicOneEvent {
             Err(err) => Err(err.into()),
         }
     }
+
+    /// Splits the evtents into deliverable and non-deliverable events. Non-deliverable events are checked against the database to see if they can be delivered.
+    /// If `require_history` is true, an error is returned if any event is missing history, oherwise they will not be marked as deliverable.
+    async fn with_deliverable_marker<'a>(
+        events: &'a [EventInsertable],
+        pool: &SqlitePool,
+        require_history: bool,
+    ) -> Result<Vec<(bool, &'a EventInsertable)>> {
+        // move all the init events to the front so we make sure to add them first and get the deliverable order correct
+        let mut insert = Vec::with_capacity(events.len());
+        let mut insert_check_history = Vec::with_capacity(events.len());
+        for event in events {
+            if event.body.event_type() == EventType::Init {
+                insert.push((true, event));
+            } else {
+                insert_check_history.push(event);
+            }
+        }
+
+        for event in insert_check_history {
+            match &event.body.header {
+                EventHeader::Init { .. } => {
+                    unreachable!("Init events should have been filtered out")
+                }
+                EventHeader::Data { cid, prev, .. } | EventHeader::Time { cid, prev, .. } => {
+                    // check for prev in this set and fallback to database
```

Comment:
Hmm, I think you're right. Good catch! I should probably write a better test and make sure 😄. This would fail if we somehow got 3 writes from the API for a single stream out of order. For recon discovered events, we would miss marking them immediately (but wouldn't error), and the stream would get sent to the task and it would find them, but it shouldn't need to.